crates/redeem/src/lib.rs

```diff
@@ -462,38 +521,94 @@ mod self_redeem {
             &vault_id.account_id,
         )?;
 
-        Pallet::<T>::release_replace_collateral(vault_id, consumed_issued_tokens)?;
         Ok(())
     }
 }
 // "Internal" functions, callable by code.
 #[cfg_attr(test, mockable)]
 impl<T: Config> Pallet<T> {
+    fn _request_replace(
+        old_vault_account: T::AccountId,
+        currency_pair: DefaultVaultCurrencyPair<T>,
+        amount: BalanceOf<T>,
+        new_vault_id: DefaultVaultId<T>,
+        griefing_currency: CurrencyId<T>,
+    ) -> Result<(), DispatchError> {
+        // Create VaultId for the old vault
+        let old_vault_id = VaultId::new(
+            old_vault_account.clone(),
+            currency_pair.collateral,
+            currency_pair.wrapped,
+        );
+
+        // Define old and new vault references
+        let old_vault = AccountOrVault::Vault(old_vault_id.clone());
+        let new_vault = AccountOrVault::Vault(new_vault_id.clone());
+
+        // probably this check is not strictly required, but it's better to give an
+        // explicit error rather than insufficient balance
+        ensure!(
+            old_vault_id.wrapped_currency() == new_vault_id.wrapped_currency(),
+            Error::<T>::InvalidWrappedCurrency
+        );
+
+        // don't allow vaults to replace themselves
+        ensure!(old_vault != new_vault, Error::<T>::ReplaceSelfNotAllowed);
+
+        // Calculate requestable tokens for old vault
+        let max_requestable_tokens: Amount<T> =
+            ext::vault_registry::requestable_to_be_replaced_tokens::<T>(&old_vault_id)?;
+        let requestable_tokens = amount.min(max_requestable_tokens.amount());
+
+        // Initiate issue call
+        let (issue_id, btc_address) = ext::issue::request_vault_issue::<T>(
+            old_vault.clone(),
+            requestable_tokens,
+            new_vault_id.clone(),
+            griefing_currency,
+        )?;
+
+        // Initiate redeem call
+        let redeem_id =
+            Self::_request_redeem(new_vault, requestable_tokens, btc_address, old_vault_id, Some(issue_id))?;
+
+        // Store the request IDs
+        Self::insert_request(&redeem_id, &issue_id);
+
+        Ok(())
+    }
+
     fn _request_redeem(
-        redeemer: T::AccountId,
+        redeemer: AccountOrVault<T::AccountId, T::CurrencyId>,
         amount_wrapped: BalanceOf<T>,
         btc_address: BtcAddress,
         vault_id: DefaultVaultId<T>,
+        issue_id: Option<H256>,
     ) -> Result<H256, DispatchError> {
         let amount_wrapped = Amount::new(amount_wrapped, vault_id.wrapped_currency());
 
-        let redeemer_balance = ext::treasury::get_balance::<T>(&redeemer, vault_id.wrapped_currency());
-        ensure!(
-            amount_wrapped.le(&redeemer_balance)?,
-            Error::<T>::AmountExceedsUserBalance
-        );
+        // if redeemer is vault then it's a replace request, skip wrapped tokens balance check for
+        // it
+        if !redeemer.is_vault_account() {
```

Comment:
This check is being done three times in this function. Is it possible to refactor to reduce this to a single check so that it's easier to see the difference in behavior between vault & non-vault code paths?